mobile/ios/ios-device.ts

```diff
@@ -320,11 +333,135 @@ export class IOSDevice implements Mobile.IIOSDevice {
+                localSocket.on('end', () => {
+                    console.log('Localhost client disconnected!');
+                    process.exit(0);
+                });
+
+				deviceSocket = this.connectToPort(port);
+				console.log("Connected localhost 8080 to device 8080");
+
+				buff.forEach((data) => {
+					deviceSocket.write(data);
+				});
+
+				deviceSocket.on('data', (data) => {
+					localSocket.write(data);
+				});
+
+				deviceSocket.on('end', (data) => {
+					console.log("Device socket closed!");
+					process.exit(0);
+				});
+
+			});
+
+            localServer.listen(port, () => {
+                console.log("Opened localhost 8080");
+            });
+        });
+    }
+
+    private printHowToTerminate() {
+        console.log("Setting up debugger proxy...");
+        console.log();
+        console.log("Press Ctrl + C to terminate!");
+        console.log();
+    }
+
+    private postDebugNotification(packageName: string, notification: string, notificationProxyClient: iOSProxyServices.NotificationProxyClient): void {
+        var attachRequestMessage = packageName + ":NativeScript.Debug.AttachRequest";
+        notificationProxyClient.postNotification(attachRequestMessage);
+        console.log("Send " + notification);
+    }
+
+    private attachAtStartup(packageName: string) {
+        this.openSafariFrontEnd(packageName);
+
+        var _injector = this.$injector;
+        var notificationProxyClient: iOSProxyServices.NotificationProxyClient = _injector.resolve(iOSProxyServices.NotificationProxyClient, { device: this });
+
+        this.printHowToTerminate();
+
+        var appLaunchMessage = packageName + ":NativeScript.Debug.AppLaunching";
+        notificationProxyClient.addObserver(appLaunchMessage, () => {
+            console.log("Got AppLaunching");
+
+            this.attachWhenReady(packageName, notificationProxyClient);
+            this.postDebugNotification(packageName, "WaitForDebugger", notificationProxyClient);
+        });
+    }
+
+    private attachAtRuntime(packageName: string): void {
+        this.openSafariFrontEnd(packageName);
+
+        var _injector = this.$injector;
+        var notificationProxyClient: iOSProxyServices.NotificationProxyClient = _injector.resolve(iOSProxyServices.NotificationProxyClient, { device: this });
+
+        this.printHowToTerminate();
+        this.attachWhenReady(packageName, notificationProxyClient);
+        this.postDebugNotification(packageName, "AttachRequest", notificationProxyClient);
+    }
+
+    private attachDebugger(packageName: string): void {
+        this.attachAtRuntime(packageName);
+    }
+
+    private detachDebugger(packageName: string): void {
+        console.log("detachDebugger: " + packageName);
+    }
+
+	private startAppWithDebugger(packageFile: string, packageName: string): void {
+        this.deploy(packageFile, packageName).wait();
+        this.attachAtStartup(packageName);
+        this.runApplication(packageName).wait();
```

Comment:
Is it a problem to call runApplication when attachAtStartup's addObserver is not finished (when we have not received  "Got AppLaunching")?